### ISSUE TITLE
Typed Handles p2

### ISSUE THREAD

**Issue by iampremo** [YoYoGames/GameMaker-Feature-Requests#1112]

**Comment by gnysek** [comment]
Seems it's about assets in asset tree.

**Comment by rwkay** [comment]
Added support for the following as typed refs.

* Buffers
* vertex formats

planning on working on
* vertex buffers
* surfaces
* shaders

**Comment by rwkay** [comment]
See https://github.com/YoYoGames/GameMaker/pull/2935 this is fixed in 2023.11 - I checked with the debugger and it seems to work as expected with buffers and surfaces....

**Comment by sihammill** [comment]
Verified in IDE v2023.1100.0.418  Runtime v2023.1100.0.438

**Comment by BackYard321** [comment]
> Added support for the following as typed refs.
> 
>     * Buffers
> 
>     * vertex formats
> 
> 
> planning on working on
> 
>     * vertex buffers
> 
>     * surfaces

Any plans for scripts/built-in functions?

**Comment by gnysek** [comment]
@BackYard321 This was already done for 2023.8

https://github.com/YoYoGames/GameMaker-Feature-Requests/issues/1105

As per runtime release notes:

> This affects DS structs, Objects, Sprites, Sounds, Rooms, Backgrounds (tilemaps), Paths, **Scripts**, Fonts, Timelines, Tiles (tilesets), Shaders, Sequences, AnimCurves, Instances and Particle Systems, so please be aware that your code may now need fixes before it will work in 2023.8 if you have actually had invalid function calls all along and old versions didn’t tell you

**Comment by BackYard321** [comment]
> @BackYard321 This was already done for 2023.8
> 
> [Typed Handles p1 - #1105](https://github.com/YoYoGames/GameMaker-Feature-Requests/issues/1105)
> 
> As per runtime release notes:
> 
> > This affects DS structs, Objects, Sprites, Sounds, Rooms, Backgrounds (tilemaps), Paths, **Scripts**, Fonts, Timelines, Tiles (tilesets), Shaders, Sequences, AnimCurves, Instances and Particle Systems, so please be aware that your code may now need fixes before it will work in 2023.8 if you have actually had invalid function calls all along and old versions didn’t tell you

It was supposed to be added, but never actually was - I discussed this in the release thread and submitted a report. If you type `msg(instance_destroy)` or `msg(userScript)`, it will give you a number, not a ref.

**Comment by rwkay** [comment]
user scripts are definitely references now (it is in 2023.11) I am not 100% sure on built in functions....

**Comment by gnysek** [comment]
@rwkay just checked in Runtime v2023.1100.0.450 :
```gml
show_debug_message(is_handle(Script1)); // 0
show_debug_message(is_handle(method(undefined, Script1))); // 0
show_debug_message(is_handle(instance_destroy)); // 0

show_debug_message(typeof(Script1)); // number
show_debug_message(typeof(instance_destroy)); // number
```
Either still a bug, or bugfix is yet not merged?

**Comment by BackYard321** [comment]
> user scripts are definitely references now (it was a bug fix, it is in 2023.11) I am not 100% sure on built in functions....

Sweet, thank you! I dug up the bug report I submitted and saw it was marked as "not planned", so I was concerned for a sec. Built-in functions would be nice too tbh, just so you could have the equivalent of `is_callable()` for actually checking whether something is a method/function ref, rather than just a number corresponding to a function.
